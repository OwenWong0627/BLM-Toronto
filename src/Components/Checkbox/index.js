import React from 'react';
import PropTypes from 'prop-types';

const Checkbox = ({ type = 'checkbox', name, value, checked, onChange }) => (
   <li><label><input type={type} name={name} checked={checked} onChange={onChange} />{value}</label></li>
);

Checkbox.propTypes = {
   type: PropTypes.string,
   name: PropTypes.string.isRequired,
   value: PropTypes.string.isRequired,
   checked: PropTypes.bool.isRequired,
   onChange: PropTypes.func.isRequired,
}

export default Checkbox;
