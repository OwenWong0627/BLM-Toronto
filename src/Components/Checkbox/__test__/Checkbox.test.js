import React from 'react';
import ReactDOM from 'react-dom';
import Checkbox from '..';

import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

const mockCheckboxes = [
   { name: "Category 1" },
   { name: "Category 2" },
   { name: "Category 3" },
   { name: "Category 4" },
   { name: "Category 5" }
];

afterEach(cleanup);
it('Renders without crashing', () => {
   const div = document.createElement("div");
   ReactDOM.render(mockCheckboxes.map(item => (
      <Checkbox
         key={item.name}
         name={item.name}
         checked={true}
         onChange={() => { }}
      />)), div);
});

describe('Renders a checkbox with the correct text content and default checked value', () => {
   it('Checks if the checkbox component has the type attribute of checkbox', () => {
      const { getByTestId } = render(<Checkbox key="Category 0" name="Category 0" checked={false} onChange={() => { }} />);
      expect(getByTestId("checkbox")).toHaveAttribute("type", "checkbox");
   })
   it('Checks if the checkbox component has the text content matching the name prop of the component', () => {
      const { getByTestId } = render(<Checkbox key="Category 0" name="Category 0" checked={false} onChange={() => { }} />);
      expect(getByTestId("checkbox-text")).toHaveTextContent("Category 0");
   });
   it('Checks if the checkbox component has the default checked value matching the boolean checked prop of the component', () => {
      const { getByTestId } = render(<Checkbox key="Category 0" name="Category 0" checked={false} onChange={() => { }} />);
      expect(getByTestId("checkbox")).not.toBeChecked();
   });
});

test('When clicked once, the onChange of the component is only executed once', () => {
   const mockFunction = jest.fn();
   const { getByTestId } = render(<Checkbox key="Category 0" name="Category 0" checked={false} onChange={mockFunction} />);
   fireEvent.click(getByTestId("checkbox"));
   expect(mockFunction.mock.calls.length).toBe(1);
});

it('matches snapshot', () => {
   const tree = renderer.create(mockCheckboxes.map(item => (
      <Checkbox
         key={item.name}
         name={item.name}
         checked={true}
         onChange={() => { }}
      />))).toJSON();
   expect(tree).toMatchSnapshot();
});