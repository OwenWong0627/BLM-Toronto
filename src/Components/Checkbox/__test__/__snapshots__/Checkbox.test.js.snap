// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <li>
    <label
      data-testid="checkbox-text"
    >
      <input
        checked={true}
        data-testid="checkbox"
        name="Category 1"
        onChange={[Function]}
        type="checkbox"
      />
      Category 1
    </label>
  </li>,
  <li>
    <label
      data-testid="checkbox-text"
    >
      <input
        checked={true}
        data-testid="checkbox"
        name="Category 2"
        onChange={[Function]}
        type="checkbox"
      />
      Category 2
    </label>
  </li>,
  <li>
    <label
      data-testid="checkbox-text"
    >
      <input
        checked={true}
        data-testid="checkbox"
        name="Category 3"
        onChange={[Function]}
        type="checkbox"
      />
      Category 3
    </label>
  </li>,
  <li>
    <label
      data-testid="checkbox-text"
    >
      <input
        checked={true}
        data-testid="checkbox"
        name="Category 4"
        onChange={[Function]}
        type="checkbox"
      />
      Category 4
    </label>
  </li>,
  <li>
    <label
      data-testid="checkbox-text"
    >
      <input
        checked={true}
        data-testid="checkbox"
        name="Category 5"
        onChange={[Function]}
        type="checkbox"
      />
      Category 5
    </label>
  </li>,
]
`;
